proposals/nnnn-attached-macros.md

```diff
@@ -0,0 +1,601 @@
+
+As with freestanding macros, attached declaration macros are declared with `macro`, and have [type-checked macro arguments](https://github.com/apple/swift-evolution/blob/main/proposals/0382-expression-macros.md#type-checked-macro-arguments-and-results) that allow their behavior to be customized. Attached macros are identified with the `@attached` attribute, which also provides the specific role as well as [any names they introduce](https://github.com/DougGregor/swift-evolution/blob/freestanding-macros/proposals/nnnn-freestanding-macros.md#up-front-declarations-of-newly-introduced-names). For example, the aforemented macro emulating property wrappers could be declared as follows:
+
+```swift
+@attached(peer, names: prefixed(_))
+@attached(accesor)
+macro wrapProperty<WrapperType>()
+```
+
+All attached macros are implemented as types that conform one of the protocols that inherits from the `AttachedMacro` protocol.  Like the [`Macro` protocol](https://github.com/apple/swift-evolution/blob/main/proposals/0382-expression-macros.md#macro-protocols), the `AttachedMacro` protocol has no requirements, but is used to organize macro implementations.
+
+```swift
+public protocol AttachedMacro: Macro { }
+```
+
+Each attached macro role will have its own protocol that inherits `AttachedMacro`.
+
+#### Peer macros
+
+Peer macros produce new declarations alongside the declaration to which they are attached.  For example, here is a declaration of a macro that introduces a completion-handler version of a given asynchronous function:
+
+```swift
+@attached(peer, names: overloaded) macro addCompletionHandler()
+```
+
+This macro uses the `@attached` attribute to indicate that it is an attached peer attribute. The `names` argument specifies that how the names of the peer declarations are created, using an extended form of the scheme introduced with [freestanding macros](https://github.com/DougGregor/swift-evolution/blob/freestanding-macros/proposals/nnnn-freestanding-macros.md#up-front-declarations-of-newly-introduced-names). In this case, `overloaded` means that this macro will produce a peer that is overloaded with the declaration to which it is attached, i.e., it has the same base name.
```

Comment:
Since functions can be distinguished by their full decl name with argument labels, what is the significance of telling the compiler that the peer declaration has the same base name? Can you expand on what the compiler does with this information?

Reply:
The compiler tends to ignore the labels when it does name lookup. I've added some more text about how this information is used.